core/src/test/scala/org/apache/spark/util/collection/ExternalAppendOnlyMapSuite.scala

```diff
@@ -414,7 +415,106 @@ class ExternalAppendOnlyMapSuite extends SparkFunSuite with LocalSparkContext {
     sc.stop()
   }
 
-  test("external aggregation updates peak execution memory") {
+  test("SPARK-22713 spill during iteration leaks internal map") {
+    val size = 1000
+    val conf = createSparkConf(loadDefaults = true)
+    sc = new SparkContext("local-cluster[1,1,1024]", "test", conf)
+    val map = createExternalMap[Int]
```

Comment:
this requires using something like [scalatest's eventually](http://doc.scalatest.org/1.8/org/scalatest/concurrent/Eventually.html), don't you think?

Reply:
`AccumulatorSuite` has a similar pattern and it works well(search System.gc). Maybe we don't need `eventually`?

Reply:
Anyway `eventually` is availabe in Spark test, we can use it if needed.

Reply:
@cloud-fan , I've tried the `System.gc` approach with and without `eventually`, and for no avail.
seems like scalatest's asserts use some black magic macro based voodoo to generate a local method/val (when I got about 10 classes deep into my quest, I gave up) 
basically what I'm seeing in visualvm is the expected ref via the `WeakReference` and another surprising one via `UnaryMacroBool.left`.

it seems this sneaky ref was generated by the following assertion: `assert(map.currentMap == null)`, but it could have easily been generated elsewhere.

@cloud-fan , @gatorsmile , can you please confirm if and how can we import the scala code? otherwise, can you think of an alternative approach for testing this?

Reply:
@cloud-fan, how do you suggest to progress with this?

Reply:
@cloud-fan , can we move on with this?

Reply:
@cloud-fan ?